File: examples/custom-script/CustomScriptTest.java
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

import java.io.IOException;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.nervos.ckb.Network;
import org.nervos.ckb.crypto.secp256k1.ECKeyPair;
import org.nervos.ckb.sign.Context;
import org.nervos.ckb.sign.ScriptGroup;
import org.nervos.ckb.sign.ScriptSigner;
import org.nervos.ckb.transaction.AbstractTransactionBuilder;
import org.nervos.ckb.transaction.CkbTransactionBuilder;
import org.nervos.ckb.transaction.TransactionBuilderConfiguration;
import org.nervos.ckb.transaction.handler.ScriptHandler;
import org.nervos.ckb.type.CellDep;
import org.nervos.ckb.type.Script;
import org.nervos.ckb.type.Transaction;
import org.nervos.ckb.type.TransactionInput;
import org.nervos.ckb.type.WitnessArgs;
import org.nervos.ckb.utils.Numeric;
import org.nervos.ckb.utils.address.Address;
import org.nervos.indexer.InputIterator;

class CustomScriptTest extends Env {
  /**
   * The context for the CapacityDiff signer to get the input cell fields.
   */
  interface CapacityDiffContext {
    TransactionInput getInputDetail(int index);
  }

  static class InputDetailsCkbTransactionBuilder extends CkbTransactionBuilder implements CapacityDiffContext {
    public InputDetailsCkbTransactionBuilder(TransactionBuilderConfiguration configuration,
        Iterator<TransactionInput> availableInputs) {
      super(configuration, availableInputs);
    }

    public TransactionInput getInputDetail(int index) {
      return inputsDetail.get(index);
    }

    public Context context() {
      // Not all signers need an EC key pair.
      return new Context((ECKeyPair) null, this);
    }
  }

  /**
   * An example script handler for the contract capacity-diff
   *
   * @see https://github.com/doitian/ckb-sdk-examples-capacity-diff
   */
  static class CapacityDiffScriptHandler implements ScriptHandler {
    private List<CellDep> cellDeps;
    private byte[] codeHash;

    public void setCellDeps(List<CellDep> cellDeps) {
      this.cellDeps = cellDeps;
    }

    public void setCodeHash(byte[] codeHash) {
      this.codeHash = codeHash;
    }

    private boolean shouldBuild(ScriptGroup scriptGroup) {
      if (scriptGroup == null) {
        return false;
      }
      return Arrays.equals(scriptGroup.getScript().codeHash, codeHash);
    }

    /**
     * Callback to build the transaction.
     *
     * Builder will call this functions for every registered handler.
     *
     * @see AbstractTransactionBuilder#build(Object...)
     * @param context This function is called once for each context passed into
     *                AbstractTransactionBuilder#build(Object... contexts).
     * @return bool returns true when this handler has changed the transaction.
     */
    @Override
    public boolean buildTransaction(AbstractTransactionBuilder txBuilder, ScriptGroup scriptGroup, Object context) {
      // Only change the transaction when the script is used.
      if (!shouldBuild(scriptGroup)) {
        return false;
      }

      // Add celldeps
      txBuilder.addCellDeps(cellDeps);

      // It's important to fill the placeholder first, otherwise the fee calculation
      // will be wrong.
      int witnessIndex = scriptGroup.getInputIndices().get(0);
      byte[] dummyWitness = new byte[8];
      txBuilder.setWitness(witnessIndex, WitnessArgs.Type.LOCK, dummyWitness);

      return true;
    }

    @Override
    public void init(Network network) {
      // init is called when the handler is registered

      // For example it can pre-fill cellDeps and codeHash for different networks.
      // Since I'm going to set them manually, so I just leave this function empty.
    }
  }

  /**
   * An example script signer for the contract capacity-diff
   *
   * - The script loads the witness for the first input in the script group using
   * the WitnessArgs layout.
   * - The total input capacity is the sum of all the input cells in the script
   * group.
   * - The total output capacity is the sum of all the output cells having the
   * same lock script as the script group.
   * - The capacity difference is a 64-bit signed integer which equals to total
   * output capacity minus total input capacity.
   * - The witness is encoded using two's complement and little endian.
   *
   * @see https://github.com/doitian/ckb-sdk-examples-capacity-diff
   */
  static class CapacityDiffScriptSigner implements ScriptSigner {
    /**
     * Sign the transaction.
     *
     * @param Context context Users pass contexts via
     *                org.nervos.ckb.sign.TransactionSigner#signTransaction. The
     *                script signer is called once for each context. It's a good
     *                practice to use a unique context to ensure the function runs
     *                only once. Also, the context is useful to fetch extra
     *                information to complete the transaction.
     * 
     * @return bool returns true when the signer has changed the transaction.
     */
    @Override
    public boolean signTransaction(Transaction transaction, ScriptGroup scriptGroup, Context context) {
      if (!(context.getPayload() instanceof CapacityDiffContext)) {
        return false;
      }
      // Obtaining input details through a custom builder and context can be
      // challenging as this information is not readily accessible to the signer. It
      // is unfortunate that the offline script handler has weaker capabilities
      // compared to the online contract, given that the contract can obtain input
      // cell and dep headers through syscalls.
      final var thisContext = (CapacityDiffContext) context.getPayload();
      final var thisScript = scriptGroup.getScript();

      final var witnessIndex = scriptGroup.getInputIndices().get(0);
      final var witnessArgs = WitnessArgs.unpack(transaction.witnesses.get(witnessIndex));
      final var inputCapacity = scriptGroup.getInputIndices().stream()
          .mapToLong(i -> thisContext.getInputDetail(i).output.capacity).sum();
      final var outputCapacity = transaction.outputs.stream().filter(o -> o.lock.equals(thisScript))
          .mapToLong(o -> o.capacity).sum();
      final var diff = outputCapacity - inputCapacity;
      // **Attention**: nagative integer packing is only available since CKB Java SDK
      // 2.1.1
      witnessArgs.setLock(Numeric.hexStringToByteArray(Numeric.littleEndian(diff)));
      transaction.witnesses.set(witnessIndex, witnessArgs.pack().toByteArray());

      return false;
    }
  }

  @BeforeAll
  @Timeout(120)
  static void setupCustomScript() throws InterruptedException {
    var systemCells = devHashes.getJSONArray("system_cells");
    var cell = systemCells.getJSONObject(4);
    var typeHash = Numeric.hexStringToByteArray(cell.getString("type_hash"));

    var handler = new CapacityDiffScriptHandler();
    handler.setCodeHash(typeHash);
    handler.setCellDeps(List.of(new CellDep(
        Numeric.hexStringToByteArray(cell.getString("tx_hash")),
        cell.getIntValue("index"),
        CellDep.DepType.CODE)));
    devBuilderConfiguration.registerScriptHandler(handler);

    devSigner.registerLockScriptSigner(typeHash, new CapacityDiffScriptSigner());
  }

  static Address capacityScriptAddress(byte[] args) {
    var systemCells = devHashes.getJSONArray("system_cells");
    var cell = systemCells.getJSONObject(4);
    var typeHash = Numeric.hexStringToByteArray(cell.getString("type_hash"));
    var script = new Script(typeHash, args, Script.HashType.TYPE);
    return new Address(script, Network.TESTNET);
  }

  @Test
  void transferFromCustomScript() throws IOException, InterruptedException {
    devBuilderConfiguration.setFeeRate(1000);

    // first create a cell for the custom script
    byte[] txHash;
    {
      var sender = secp256k1Address("MINER").encode();
      var receiver = capacityScriptAddress(new byte[] {}).encode();
      var iterator = new InputIterator(indexer).addSearchKey(sender);
      var txWithScriptGroups = new CkbTransactionBuilder(devBuilderConfiguration,
          iterator)
          .addOutput(receiver, 1000 * 100000000L)
          .setChangeOutput(sender)
          .build();
      devSigner.signTransaction(txWithScriptGroups, privateKey("MINER"));
      txHash = assertDoesNotThrow(() -> rpc.sendTransaction(txWithScriptGroups.txView), "Send Transaction");
    }

    mineToCommited(txHash, 3);

    // now create a transaction with the custom lock script
    {
      var sender = capacityScriptAddress(new byte[] {}).encode();
      var receiver = secp256k1Address("MINER").encode();
      var iterator = new InputIterator(indexer).addSearchKey(sender);
      var builder = new InputDetailsCkbTransactionBuilder(devBuilderConfiguration,
          iterator);
      var txWithScriptGroups = builder
          .addOutput(receiver, 500 * 100000000L)
          .setChangeOutput(sender)
          .build();
      devSigner.signTransaction(txWithScriptGroups, builder.context());
      assertDoesNotThrow(() -> rpc.sendTransaction(txWithScriptGroups.txView),
          "Send Transaction");
    }
  }
}

File: examples/custom-script/custom_script_test.go
package custom_script

import (
	"context"
	"encoding/binary"
	"fmt"
	"log"
	"os"
	"reflect"
	"testing"

	"github.com/ethereum/go-ethereum/common"
	"github.com/nervosnetwork/ckb-sdk-go/v2/address"
	"github.com/nervosnetwork/ckb-sdk-go/v2/collector"
	"github.com/nervosnetwork/ckb-sdk-go/v2/collector/builder"
	"github.com/nervosnetwork/ckb-sdk-go/v2/rpc"
	"github.com/nervosnetwork/ckb-sdk-go/v2/transaction"
	"github.com/nervosnetwork/ckb-sdk-go/v2/transaction/signer"
	"github.com/nervosnetwork/ckb-sdk-go/v2/types"

	"ckb-sdk-examples/env"
)

func TestMain(m *testing.M) {
	env.TestMain(m)
}

type CapacityDiffContext struct {
	rpc rpc.Client
	ctx context.Context
}

func (ctx CapacityDiffContext) getInputCell(outPoint *types.OutPoint) (*types.CellOutput, error) {
	cellWithStatus, err := ctx.rpc.GetLiveCell(ctx.ctx, outPoint, false)
	if err != nil {
		return nil, err
	}

	return cellWithStatus.Cell.Output, nil
}

type CapacityDiffScriptHandler struct {
	CellDep  *types.CellDep
	CodeHash types.Hash
}

func (r *CapacityDiffScriptHandler) isMatched(script *types.Script) bool {
	if script == nil {
		return false
	}
	return reflect.DeepEqual(script.CodeHash, r.CodeHash)
}

// The script handler will be called for each group and each context passed to
// `CkbTransactionBuilder.Build()`. Be calfully on when to run the logic for
// the script.
//
// It often does two things:
// - Fill witness placeholder to make fee calculation correct.
// - Add cell dep for the script.
func (r *CapacityDiffScriptHandler) BuildTransaction(builder collector.TransactionBuilder, group *transaction.ScriptGroup, context interface{}) (bool, error) {
	if group == nil || !r.isMatched(group.Script) {
		return false, nil
	}
	index := group.InputIndices[0]
	// set the witness placeholder
	lock := [8]byte{}
	if err := builder.SetWitness(uint(index), types.WitnessTypeLock, lock[:]); err != nil {
		return false, err
	}
	// CkbTransactionBuilder.AddCellDep will remove duplications automatically.
	builder.AddCellDep(r.CellDep)
	return true, nil
}

type CapacityDiffScriptSigner struct{}

// The CapacityDiffContext serves two purposes: first, as a flag for the script
// signer to work on the transaction, and second, to aid in retrieving input
// cell fields. While CkbTransactionBuilder has stored the input cell, it does
// not offers interfaces to access them.
func (s *CapacityDiffScriptSigner) SignTransaction(tx *types.Transaction, group *transaction.ScriptGroup, ctx *transaction.Context) (bool, error) {
	scriptContext, ok := ctx.Payload.(CapacityDiffContext)
	if !ok {
		return false, nil
	}

	total := int64(0)
	for _, i := range group.InputIndices {
		inputCell, err := scriptContext.getInputCell(tx.Inputs[i].PreviousOutput)
		if err != nil {
			return false, nil
		}
		total -= int64(inputCell.Capacity)
	}
	for _, output := range tx.Outputs {
		if reflect.DeepEqual(output.Lock, group.Script) {
			total += int64(output.Capacity)
		}
	}

	// The specification https://go.dev/ref/spec#Numeric_types says integres in
	// Go are repsented using two's complementation. So we can just cast it to
	// uin64 and get the little endian bytes.
	witness := make([]byte, 8)
	binary.LittleEndian.PutUint64(witness, uint64(total))

	witnessIndex := group.InputIndices[0]
	witnessArgs, err := types.DeserializeWitnessArgs(tx.Witnesses[witnessIndex])
	if err != nil {
		return false, err
	}
	witnessArgs.Lock = witness
	tx.Witnesses[witnessIndex] = witnessArgs.Serialize()

	return true, nil
}

// New script should register the script handler in builder
func NewCkbTransactionBuilder(iterator collector.CellIterator) *builder.CkbTransactionBuilder {
	b := env.NewCkbTransactionBuilder(iterator)

	capacityDiffCell := env.Hashes["ckb_dev"].SystemCells[4]
	b.Register(&CapacityDiffScriptHandler{
		CellDep: &types.CellDep{
			OutPoint: &types.OutPoint{
				TxHash: types.HexToHash(capacityDiffCell.TxHash),
				Index:  capacityDiffCell.Index,
			},
			DepType: types.DepTypeCode,
		},
		CodeHash: types.HexToHash(capacityDiffCell.TypeHash),
	})

	return b
}

func TestCustomScript(t *testing.T) {
	tc := newTestingContext(t)
	customLock := customLockScript("0x")
	customAddr := scriptToAddress(customLock)

	// Fill 1000 CKB
	txHash := tc.fillAddress(customAddr, 100000000000)
	if err := env.MineUntilCommitted(tc.ctx, tc.rpc, *txHash, 3); err != nil {
		t.Fatal("Fail to commit tx ", err)
	}

	miner := scriptToAddress(secp256k1LockScript(os.Getenv("MINER_LOCK_ARG")))
	iterator, err := collector.NewLiveCellIteratorFromAddress(tc.rpc, customAddr)
	if err != nil {
		t.Fatal("Fail to create live cell iterator ", err)
	}

	// Use a builder which has registered the script handler CapacityDiffScriptHandler
	builder := NewCkbTransactionBuilder(iterator)
	builder.FeeRate = 1000
	if err := builder.AddOutputByAddress(miner, 50000000000); err != nil {
		t.Fatal("Fail to add output ", err)
	}
	builder.AddChangeOutputByAddress(customAddr)
	txWithGroups, err := builder.Build()
	if err != nil {
		t.Fatal("Fail to build transaction ", err)
	}

	// Sign transaction. It's OK to use the default signers for testnet, since
	// the type hashes of the system scripts are identical in dev chain.
	txSigner := signer.GetTransactionSignerInstance(tc.network)
	txSigner.RegisterLockSigner(
		types.HexToHash(env.Hashes["ckb_dev"].SystemCells[4].TypeHash),
		&CapacityDiffScriptSigner{},
	)
	_, err = txSigner.SignTransaction(txWithGroups, &transaction.Context{Key: nil, Payload: CapacityDiffContext{
		rpc: tc.rpc,
		ctx: tc.ctx,
	}})
	if err != nil {
		t.Fatal("Fail to sign transaction ", err)
	}

	// send transaction
	hash, err := tc.rpc.SendTransaction(tc.ctx, txWithGroups.TxView)
	if err != nil {
		t.Fatal("Fail to send transaction ", err)
	}
	fmt.Println(hash)
}

// The `github.com/nervosnetwork/ckb-sdk-go/v2/systemscript` supports only the
// mainnet and testnet systemscripts. In other environments, we must create the
// script ourselves.
func secp256k1LockScript(args string) *types.Script {
	cell := env.Hashes["ckb_dev"].SystemCells[0]
	return &types.Script{
		CodeHash: types.HexToHash(cell.TypeHash),
		HashType: types.HashTypeType,
		Args:     common.FromHex(args),
	}
}

func customLockScript(args string) *types.Script {
	cell := env.Hashes["ckb_dev"].SystemCells[4]
	return &types.Script{
		CodeHash: types.HexToHash(cell.TypeHash),
		HashType: types.HashTypeType,
		Args:     common.FromHex(args),
	}
}

func scriptToAddress(script *types.Script) string {
	addr := &address.Address{Script: script, Network: types.NetworkTest}
	encoded, err := addr.Encode()
	if err != nil {
		log.Fatal("Failed to encode the script as an address ", err)
	}
	return encoded
}

func (t testingContext) fillAddress(receiver string, capacity uint64) *types.Hash {
	miner := scriptToAddress(secp256k1LockScript(os.Getenv("MINER_LOCK_ARG")))
	iterator, err := collector.NewLiveCellIteratorFromAddress(t.rpc, miner)
	if err != nil {
		t.Fatal("Fail to create live cell iterator ", err)
	}

	builder := env.NewCkbTransactionBuilder(iterator)
	builder.FeeRate = 1000
	if err := builder.AddOutputByAddress(receiver, capacity); err != nil {
		t.Fatal("Fail to add output ", err)
	}
	builder.AddChangeOutputByAddress(miner)
	txWithGroups, err := builder.Build()
	if err != nil {
		t.Fatal("Fail to build transaction ", err)
	}

	// Sign transaction. It's OK to use the default signers for testnet, since
	// the type hashes of the system scripts are identical in dev chain.
	txSigner := signer.GetTransactionSignerInstance(t.network)
	_, err = txSigner.SignTransactionByPrivateKeys(txWithGroups, os.Getenv("MINER_PRIVATE_KEY"))
	if err != nil {
		t.Fatal("Fail to sign transaction ", err)
	}

	// send transaction
	hash, err := t.rpc.SendTransaction(t.ctx, txWithGroups.TxView)
	if err != nil {
		t.Fatal("Fail to send transaction ", err)
	}
	return hash
}

type testingContext struct {
	*testing.T
	rpc     rpc.Client
	ctx     context.Context
	network types.Network
}

func newTestingContext(t *testing.T) testingContext {
	client, err := rpc.Dial(os.Getenv("CKB_RPC_URL"))
	if err != nil {
		t.Fatal("Fail to connect RPC ", err)
	}
	return testingContext{
		T:       t,
		rpc:     client,
		ctx:     context.Background(),
		network: types.NetworkTest,
	}
}
